state/exchange/actions.ts

```diff
@@ -0,0 +1,417 @@
+		const { tokensMap, tokenList } = await sdk.exchange.getOneInchTokens();
+
+		return { synthsMap, tokensMap, tokenList };
+	}
+);
+
+export const resetCurrencyKeys = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/resetCurrencyKeys',
+	async (_, { getState, dispatch, extra: { sdk } }) => {
+		const {
+			exchange: { quoteCurrencyKey, baseCurrencyKey, quoteAmount, baseAmount },
+			wallet: { walletAddress },
+		} = getState();
+
+		let baseFeeRate = undefined;
+		let rate = undefined;
+		let exchangeFeeRate = undefined;
+		let quotePriceRate = undefined;
+		let basePriceRate = undefined;
+		let txProvider = undefined;
+		let approvalStatus = undefined;
+
+		if (walletAddress) {
+			const quoteBalance = !!quoteCurrencyKey
+				? await sdk.exchange.getBalance(quoteCurrencyKey)
+				: undefined;
+
+			const baseBalance = !!baseCurrencyKey
+				? await sdk.exchange.getBalance(baseCurrencyKey)
+				: undefined;
+
+			dispatch({
+				type: 'exchange/setBalances',
+				payload: { quoteBalance: quoteBalance?.toString(), baseBalance: baseBalance?.toString() },
+			});
+
+			if (quoteCurrencyKey && baseCurrencyKey) {
+				[baseFeeRate, rate, exchangeFeeRate, quotePriceRate, basePriceRate] = await Promise.all([
+					sdk.exchange.getBaseFeeRate(baseCurrencyKey, quoteCurrencyKey),
+					sdk.exchange.getRate(baseCurrencyKey, quoteCurrencyKey),
+					sdk.exchange.getExchangeFeeRate(quoteCurrencyKey, baseCurrencyKey),
+					sdk.exchange.getQuotePriceRate(baseCurrencyKey, quoteCurrencyKey),
+					sdk.exchange.getBasePriceRate(baseCurrencyKey, quoteCurrencyKey),
+				]);
+
+				txProvider = sdk.exchange.getTxProvider(baseCurrencyKey, quoteCurrencyKey);
+
+				if (txProvider === 'synthetix') {
+					if (!!quoteCurrencyKey && !!quoteAmount && !!baseCurrencyKey) {
+						const baseAmountNoFee = wei(quoteAmount).mul(rate ?? 0);
+						const fee = baseAmountNoFee.mul(exchangeFeeRate ?? 0);
+
+						dispatch({
+							type: 'exchange/setBaseAmountRaw',
+							payload: truncateNumbers(baseAmountNoFee.sub(fee), DEFAULT_CRYPTO_DECIMALS),
+						});
+					}
+
+					if (!!baseCurrencyKey && !!baseAmount && !!quoteCurrencyKey) {
+						const inverseRate = wei(rate ?? 0).gt(0) ? wei(1).div(rate) : wei(0);
+						const quoteAmountNoFee = wei(baseAmount).mul(inverseRate);
+						const fee = quoteAmountNoFee.mul(exchangeFeeRate ?? 0);
+
+						dispatch({
+							type: 'exchange/setQuoteAmountRaw',
+							payload: truncateNumbers(quoteAmountNoFee.sub(fee), DEFAULT_CRYPTO_DECIMALS),
+						});
+					}
+				}
+
+				const needsApproval = sdk.exchange.checkNeedsApproval(baseCurrencyKey, quoteCurrencyKey);
+
+				if (needsApproval) {
+					// TODO: Handle case where allowance is not MaxUint256.
+					// Simplest way to do this is to return the allowance from
+					// checkAllowance, store it in state to do the comparison there.
+					const isApproved = await sdk.exchange.checkAllowance(
+						quoteCurrencyKey,
+						baseCurrencyKey,
+						'0'
```

Comment:
Yea as your comment above we shouldn't have approved being set bigger than 0, it should be checked against amount when making a swap, let's do that in this PR otherwise it's buggy logic.